src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.State.cs

```diff
@@ -0,0 +1,296 @@
+                var folderPath = Path.Combine(document.Folders.ToArray());
+
+                var absoluteDircetoryPath = PathUtilities.GetDirectoryName(document.FilePath);
+                var logicalDirectoryPath = PathUtilities.CombineAbsoluteAndRelativePaths(projectRoot, folderPath);
+
+                if (PathUtilities.PathsEqual(absoluteDircetoryPath, logicalDirectoryPath))
+                {
+                    allDocumentIds = linkedDocumentids.Add(document.Id);
+                    return true;
+                }
+                else
+                {
+                    allDocumentIds = default;
+                    return false;
+                }
+            }
+
+            private static string GetDefaultNamespace(ImmutableArray<Document> documents, ISyntaxFactsService syntaxFacts)
+            {
+                // For all projects containing all the linked documents, bail if 
+                // 1. Any of them doesn't have default namespace, or
+                // 2. Multiple default namespace are found. (this might be possible by tweaking project file).
+                // The refactoring depends on a single default namespace to operate.
+                var defaultNamespaceFromProjects = new HashSet<string>(
+                    documents.Select(d => d.Project.DefaultNamespace),
+                    syntaxFacts.StringComparer);
+
+                if (defaultNamespaceFromProjects.Count != 1
+                    || defaultNamespaceFromProjects.First() == null)
+                {
+                    return default;
+                }
+
+                return defaultNamespaceFromProjects.Single();
+            }
+
+            private static async Task<(bool shouldTrigger, string declaredNamespace)> TryGetNamespaceDeclarationAsync(
+                TextSpan textSpan,
+                ImmutableArray<Document> documents,
+                AbstractSyncNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax, TMemberDeclarationSyntax> service,
+                CancellationToken cancellationToken)
+            {
+                // If the cursor location doesn't meet the requirement to trigger the refactoring in any of the documents 
+                // (See `ShouldPositionTriggerRefactoringAsync`), or we are getting different namespace declarations among 
+                // those documents, then we know we can't make a proper code change. We will return false and the refactoring 
+                // will then bail. We use span of namespace declaration found in each document to decide if they are identical.
+                var spansForNamespaceDeclaration = new Dictionary<TextSpan, TNamespaceDeclarationSyntax>();
+                foreach (var document in documents)
+                {
+                    var compilationUnitOrNamespaceDeclOpt = await service.ShouldPositionTriggerRefactoringAsync(document, textSpan.Start, cancellationToken)
```

Comment:
why this refactoring specific code in the service?

Reply:
there was another one get return members. but that seems not specific to refactoring but this is really specific to refactoring.

Reply:
I don't have a chance to refactor this yet, since the service is for "sync namespace" only at the moment. I'm planning to clean this up when implementing "change namespace"